src/storage/mod.rs

```diff
@@ -1468,7 +1478,12 @@ impl<E: Engine, L: LockManager, F: KvFormat> Storage<E, L, F> {
                         .map_err(txn::Error::from);
                     statistics.add(&reader.statistics);
                     let (kv_pairs, _) = result?;
-                    let mut locks = Vec::with_capacity(kv_pairs.len());
+                    let mut locks = Vec::with_capacity(kv_pairs.len() + memory_lock_kv_pairs.len());
+                    for (key, lock) in memory_lock_kv_pairs {
+                        let lock_info =
+                            lock.into_lock_info(key.into_raw().map_err(txn::Error::from)?);
+                        locks.push(lock_info);
+                    }
```

Comment:
Sadly we cannot merge the two results by simply concatenating them. TiDB's GCWorker will assume that when a scan_lock invocation returns a (non-empty) list of locks, then there won't be any other unknown stale locks in range `[start_key, locks[len(locks)-1].key]`. So if the count of locks exceeds the `limit` param of the request, TiDB can then perform the next scan with start_key =  `locks[len(locks-1)].key`.
In other words, scan_locks should guarantee that it returns the minimal N locks in the range where N = min(number  of locks in range, limit).

Another problem is that TiDB determines whether the scanning on the range is finished by checking whether the returned locks is less than limit. So I think it's a bad idea to return more than `limit` locks. Here the code seems 
 to possibly return at most 2*limit locks (memlock*limit + persistedlock*limit). Also, it's confusing not to follow how it's actually requested by specifying `limit`.

I think you can simply sort the vector `locks` and keep only the first `limit` items if it's too many.
The GCWorker doesn't rely on the order of locks in the response (if I remember it correctly), but I still suggest to keep the order to keep the behavior consistent.

Reply:
Thanks for the reminding! The lock result merging logic is added, the `merge_by` function is used to merge the two sorted vectors, PTAL.